### PR TITLE
Fix parsing context for lambdas and modules

### PR DESCRIPTION
**Pull request by Morriar**
### Motivation

This pull-request fixes the context used when entering classes, modules and lambdas (port of https://github.com/whitequark/parser/pull/709).

This fixes a bug where the parser [rejected valid code](https://sorbet.run/#%23%20typed%3A%20true%0A%0Aclass%20A%3B%20-%3E(a%20%3D%20(return%3B%201))%20%7B%7D%3B%20end):

```ruby
class A; ->(a = (return; 1)) {}; end
#                ^^^^^^ error: invalid return in class/module body
```

Which is now correctly parsed.

### Test plan

See included automated tests.

### CODE REVIEW

File: test/whitequark/test_context_lambda.rb [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# typed: strict
+
+module A; ->(a = (return; 1)) {}; end
+class B; ->(a = (return; 1)) {}; end
```
Comment (jez) [inline test/whitequark/test_context_lambda.rb:4]: The whitequark PR also mentioned another case where the `module` was inside the lambda. I guess maybe that test doesn't make sense to add until Sorbet supports `_1` syntax?

File: third_party/parser/cc/grammars/typedruby.ypp [hunk 2]
```diff
@@ -1955,6 +1961,7 @@ opt_block_args_tail:
                     }
                   f_larglist
                     {
+                      driver.lex.context.pop();
                       driver.lex.cmdarg.push(false);
                     }
                   lambda_body
```
Comment (jez) [inline third_party/parser/cc/grammars/typedruby.ypp:1967]: https://github.com/whitequark/parser/pull/709/files#diff-7a330c23c558160f33721dcac766f436L2096-R2105

^ why doesn't the Sorbet change have something that looks like these lines?

Comment (jez) [reply third_party/parser/cc/grammars/typedruby.ypp:1967]: actually i'm having a hard time matching this PR up with the original whitequark PR. it's probably fine.

File: test/whitequark/test_context_lambda.rb (hunk 1 above)
Comment (jez) [inline test/whitequark/test_context_lambda.rb:4]: Did you confirm that this test did something different / incorrect before this change?

**Review by jez — APPROVED** [review]
This mostly looks good to me. I'm happy to merge after you have a chance to answer my questions.

File: test/whitequark/test_context_lambda.rb (hunk 1 above)
Comment (Morriar) [reply test/whitequark/test_context_lambda.rb:4]: The link given in the introduction shows the behavior with current Sorbet: https://sorbet.run/#%23%20typed%3A%20true%0A%0Aclass%20A%3B%20-%3E(a%20%3D%20(return%3B%201))%20%7B%7D%3B%20end

File: third_party/parser/cc/grammars/typedruby.ypp (hunk 2 above)
Comment (Morriar) [reply third_party/parser/cc/grammars/typedruby.ypp:1967]: Wouldn't that be that change: https://github.com/sorbet/sorbet/pull/3363/files#diff-7be0b959d201832b58982ab17587c439L1487-R1491

File: test/whitequark/test_context_lambda.rb (hunk 1 above)
Comment (Morriar) [reply test/whitequark/test_context_lambda.rb:4]: The case is currently accepted by Sorbet: https://sorbet.run/#%23%20typed%3A%20true%0A%0A-%3E(a)%20%7B%20module%20M%3B%20%22%22%3B%20end%20%7D

We need this one to be merged before #3344 so I can port the fix.

**Comment by Morriar** [comment]
Let's merge this one as this and complete the work in #3344.

**Pull request merged**